### PR TITLE
ref(analytics): Analytics Refactor + Types

### PR DESCRIPTION
**Pull request by mgaeta**
This PR refreshes the analytics modules with a few things:
- splitting up large files
- dataclasses
- python types
If there's too much going on here I think I can pull types into a separate PR.

### CODE REVIEW

File: src/sentry/analytics/events/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
 from sentry.utils.imports import import_submodules
 
-import_submodules(globals(), __name__, __path__)
+path = __path__  # type: ignore
```
Comment (evanh) [inline src/sentry/analytics/events/__init__.py:3]: Why did you need this change?

File: src/sentry/analytics/map.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+        This will hard error in some situations if you're passing a bad type
+        (like an int).
+        """
+        if value is None:
+            return value
+
+        if isinstance(value, dict):
+            # Ensure we don't mutate the original. We do not need to deepcopy;
+            # if it recurses into another Map it will once again copy itself.
```
Comment (evanh) [inline src/sentry/analytics/map.py:29]: You're aware that if the dictionary has a list or some other complex object in it, that object will not be copied? The comment is a little unclear here, I can't tell if you're saying `copy` will recurse (which it won't) or `extract` recurses.

**Review by evanh — COMMENTED** [review]
Just a couple small things. I'll leave approval to someone who understands better the semantic changes here.

File: src/sentry/analytics/map.py (hunk 2 above)
Comment (mgaeta) [reply src/sentry/analytics/map.py:29]: I just copied this class to a file so I don't know what the intention was for this comment. I'll leave a note in the code here.

File: src/sentry/analytics/events/__init__.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/analytics/events/__init__.py]: Without the `ignore`, mypy says: 
```
error: Name '__path__' is not defined
```

File: src/sentry/analytics/attribute.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any
+
+
+@dataclass
+class Attribute:
+    name: str
+    type: type = str
```
Comment (leeandher) [inline src/sentry/analytics/attribute.py:10]: Not a change, just a question for my understanding:

What does this line mean? The line above looks like it's saying the `name` property is a string, and below it, `required` property is a boolean that defaults to `True`. I'm not sure what this line means, nor how to google it 😵‍💫

File: src/sentry/analytics/base.py [hunk 4]
```diff
@@ -1,22 +1,24 @@
...
+from typing import Any
 
 from sentry.analytics.event import Event
 from sentry.utils.services import Service
 
 from .event_manager import default_manager
 
 
-class Analytics(Service):
+class Analytics(Service, abc.ABC):  # type: ignore
```
Comment (leeandher) [inline src/sentry/analytics/base.py:14]: Sorry if these questions are all simple/obvious, what does this module do? (`abc.ABC`) I don't see it used in the class.

File: src/sentry/analytics/event.py [hunk 5]
```diff
@@ -1,99 +1,50 @@
...
+    def _get_type(self, _type: Any | None = None) -> Any:
+        """
+        The Event's `type` can either be passed in as a parameter or set as a
+        property on a child class.
+        """
+        if _type is not None:
+            return _type
 
         if self.type is None:
             raise ValueError("Event is missing type")
```
Comment (leeandher) [inline src/sentry/analytics/event.py:43]: Is `NotImplementedError` more accurate here? I know you didn't make this change but just a thought looking at your comment on L20.

File: src/sentry/analytics/event_manager.py [hunk 6]
```diff
@@ -1,20 +1,22 @@
...
 
 
 class EventManager:
-    def __init__(self):
-        self._event_types = {}
+    def __init__(self) -> None:
+        self._event_types: MutableMapping[Any, Type[Event]] = {}
 
-    def register(self, event_cls: Event) -> None:
+    def register(self, event_cls: Type[Event]) -> None:
```
Comment (leeandher) [inline src/sentry/analytics/event_manager.py:12]: Another question: How does `: Type[Event]` vary from `: Event` and should we be using `Type[Event]` from now on?

**Review by leeandher — COMMENTED** [review]
Lots of questions (just for my understanding) before I approve, looks good though!

File: src/sentry/analytics/base.py (hunk 4 above)
Comment (mgaeta) [reply src/sentry/analytics/base.py:14]: `ABC` stands for "abstract base class". It marks a class as abstract meaning that you should not initialize it and that it does not have to implement all of its parents' abstract methods.
https://docs.python.org/3/library/abc.html

File: src/sentry/analytics/attribute.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/analytics/attribute.py:10]: > `type: type = str`
This line defines a member variable for the `Attribute` class. (This is pretty meta so might not be a good example for learning `dataclasses`.) I'll break it down:
- `type` - the name of  variable, literally the letters "t-y-p-e"
- `type` - the _type_ of  variable. In other variable definitions there are `int` and `str` here but in this case, this variable wants to store a _type_.
- `str` - the default value. So we know that the variable will be holding a `type`, this here tells it that if no type is passed in the default type should be "string".

File: src/sentry/analytics/base.py (hunk 4 above)
Comment (mgaeta) [reply src/sentry/analytics/base.py:14]: Actually, if we're trying to be pure, an abstract class should not be able to have an `__init__()` method. We're doing tricky things with this class were `type` is abstract-ish, meaning that it can either be passed in by parameter or inheritance. I should remove the `ABC`.

File: src/sentry/analytics/event.py (hunk 5 above)
Comment (mgaeta) [reply src/sentry/analytics/event.py:43]: I treat `raise NotImplementedError` the same as [other languages labeling an interface's method as abstract](https://en.wikipedia.org/wiki/Abstract_type). This function is reading from parameters and doing polymorphism and is just generally hacky so I'm ok with it using a different exception. Also I didn't want to change the behavior for this refactor.

File: src/sentry/analytics/event_manager.py (hunk 6 above)
Comment (mgaeta) [reply src/sentry/analytics/event_manager.py:12]:
```py
class Event:
  pass

event = Event()
type(event) == Event  # True
type(type(event)) == type(Event)  # True
type(Event)  # <class 'type'>
```
In this example, `event` is an instance and `Event` is a class. Most of our functions deal with instances because they actually contain data. The `register()` function is actually trying to use the _classes_ as a parameter so that it can register them into the `_event_types` table.

File: src/sentry/analytics/attribute.py (hunk 3 above)
Comment (leeandher) [reply src/sentry/analytics/attribute.py:10]: ahhhhh, that makes more sense now, thank you!

File: src/sentry/analytics/base.py (hunk 4 above)
Comment (leeandher) [reply src/sentry/analytics/base.py:14]: That makes a bit more sense, thanks

File: src/sentry/analytics/event.py (hunk 5 above)
Comment (leeandher) [reply src/sentry/analytics/event.py:42]: Yeah, it makes sense not to introduce new behavior in PRs like this, good call

File: src/sentry/analytics/event_manager.py (hunk 6 above)
Comment (leeandher) [reply src/sentry/analytics/event_manager.py:12]: I see, so this syntax is only used in this case because `event_cls` isn't an instance of an event, but the actual class instead. Thanks!

**Review by leeandher — APPROVED** [review]
Thanks for helping me understand the changes, looks good to me 👍

File: src/sentry/analytics/events/__init__.py (hunk 1 above)
Comment (evanh) [reply src/sentry/analytics/events/__init__.py]: Got it.
